### ISSUE TITLE
 There is no inference delay cost in the objective function of the search pattern

### ISSUE THREAD

**Issue by 66zhaofeng** [PKU-SEC-Lab/mpcvit#3]
Thank you very much for sharing the code and patiently answering any questions you may have. But I have another question:

When searching based on NAS, the objective function should include the inference delay cost（ λ Cost（ α))， In the publicly available train.py, there is no inference delay cost in the objective function of the search pattern（ λ Cost（ α))。

I also hope to receive your answers to my doubts, and I humbly request to share the **### _code_** for calculating inference delay in MPC for the purpose of verifying the results.

**Comment by zwxandy** [comment]
To enhance the flexibility of searching, we provide a way to consider the MPC situation like PolyMPCNet. The difference of the code is to change $\alpha$ to the weighted-sum of $\alpha \cdot Cost(RSAttn) + (1 - \alpha) \cdot Cost(ScaleAttn)$ when computing the regularization term. In our experiment, we find using the L1-norm of $\alpha$ for NAS can achieve the Pareto front compared with the efficient ViT baselines, so the cost term can be seen as a flexible choice for NAS.

**Comment by 66zhaofeng** [comment]
Thank you very much for your patient response.

1) Firstly, based on your feedback, I understand that in the existing train.py script:

`reg_loss += torch.norm(param, p=1)`

should be changed to:
```
weighted_sum = param * cost_rs_attn + (1 - param) * cost_scale_attn
alpha_list.append(weighted_sum.data.cpu().numpy())
```
I hope to receive your confirmation on this understanding.

2) Secondly, I have another question. The paper mentions that compared to the baseline, MPCViT latency has been reduced by 6.2 times, 2.9 times, and 1.9 times, respectively. I am curious about how the latency of MPCViT and various modules of ViT is tested in the MPC setting. 

Is there any relevant code that can be shared? I would like to request your sharing of the code for testing the latency of MPCViT and various modules of ViT in MPC.

**Comment by 66zhaofeng** [comment]
Thank you very much for your patient response.

Firstly, based on your feedback, I understand that in the existing train.py script:

&nbsp; &nbsp; &nbsp; &nbsp;reg_loss += torch.norm(param, p=1)

should be changed to:


weighted_sum = param * cost_rs_attn + (1 - param) * cost_scale_attn

alpha_list.append(weighted_sum.data.cpu().numpy())


I hope to receive your confirmation on this understanding.

Secondly,&nbsp;I have another question. The paper mentions that compared to the baseline, MPCViT latency has been reduced by 6.2 times, 2.9 times, and 1.9 times, respectively. I am curious about how the latency of MPCViT and various modules of ViT is tested in the MPC setting.

Is there any relevant code that can be shared?&nbsp;I would like to request your sharing of the code for testing the latency of MPCViT and various modules of ViT in MPC.




—
Reply to this email directly,&nbsp;view it on GitHub, or&nbsp;unsubscribe.
You are receiving this because you authored the thread.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PKU-SEC-Lab/mpcvit"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年3月24日(星期天) 晚上8:25
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PKU-SEC-Lab/mpcvit]  There is no inference delay cost in the objective function of the search pattern (Issue #3)





 
To enhance the flexibility of searching, we provide a way to consider the MPC situation like PolyMPCNet. The difference of the code is to change $\alpha$ to the weighted-sum of $\alpha \cdot Cost(RSAttn) + (1 - \alpha) \cdot Cost(ScaleAttn)$ when computing the regularization term. In our experiment, we find using the L1-norm of $\alpha$ for NAS can achieve the Pareto front compared with the efficient ViT baselines, so the cost term can be seen as a flexible choice for NAS.

**Comment by zwxandy** [comment]
> Thank you very much for your patient response.
> 
> 1. Firstly, based on your feedback, I understand that in the existing train.py script:
> 
> `reg_loss += torch.norm(param, p=1)`
> 
> should be changed to:
> 
> ```
> weighted_sum = param * cost_rs_attn + (1 - param) * cost_scale_attn
> alpha_list.append(weighted_sum.data.cpu().numpy())
> ```
> 
> I hope to receive your confirmation on this understanding.
> 
> 2. Secondly, I have another question. The paper mentions that compared to the baseline, MPCViT latency has been reduced by 6.2 times, 2.9 times, and 1.9 times, respectively. I am curious about how the latency of MPCViT and various modules of ViT is tested in the MPC setting.
> 
> Is there any relevant code that can be shared? I would like to request your sharing of the code for testing the latency of MPCViT and various modules of ViT in MPC.

For the first question, your understanding is right and the regularization term can be computed as follows.
```python
weighted_sum = param * cost_rs_attn + (1 - param) * cost_scale_attn
alpha_list.append(weighted_sum.data.cpu().numpy())
reg_loss += torch.norm(weighted_sum, p=1)
```
For the second question, we implemented the ViT model with SecretFlow framework (SPU Version: 0.3.1b0) in Jax to evaluate the inference latency. We used the semi-honest protocol SEMI-2K, and evaluated the latency under the WAN setting. Our experiments were evaluated on an Intel Xeon CPU@2.40 GHz. We used the PyTorch library for training the ViTs, and reported the model accuracy.
Since this project was completed in collaboration with the company, the Jax code is not stored on our local machine and therefore cannot be released. Thanks for your understanding and if you have any questions or need assistance, we would be happy to provide the guidance.

**Comment by 66zhaofeng** [comment]
Thank you for your answer! I have another query that requires your attention.  

During training with knowledge distillation, a parameter called --teacher ckpt/path/to/ckpt is required.  Could you please explain how you obtained the pre-training parameters for the teacher when using this parameter?  

Additionally, I would like to know whether we need to train a vit_7_4_12_teacher model separately for different datasets such as CIFAR-10/100, or if we can simply utilize existing model parameters.  

Furthermore, could you please disclose your thoughts on this and the pre trained model parameters used by the teacher?

 I would greatly appreciate it if you could address my concerns.  Thank you in advance for your assistance!

**Comment by zwxandy** [comment]
To leverage KD, our approach is to first train the teacher models, i.e., original Softmax attention ViTs on different datasets separately. And then, during training w/ KD, the teacher checkpoint is loaded to the corresponding teacher architecture. Detailed information is shown in our paper, e.g., the teacher models are trained for 600 epochs on CIFAR-10/100 and 300 epochs on Tiny-ImageNet with SGD optimizer, respectively.
Our trained teacher models is not available right now since the server storing the original parameters was broken, resulting in data loss. However, you can simply use this code to train the teacher models with original Softmax attention. 
```python
attn = (q @ k.transpose(-2, -1)) * self.scale
attn = attn.softmax(dim=-1)
```
This repo also supports to download the pre-trained ViT on CIFAR by setting `pretrained=True`; however, we did not adopt this way in our experiments. BTW, it is noteworthy that any teacher models (i.e., more powerful models) can be inserted to our framework to distill the heterogeneous ViTs.
I hope my answer can address your concerns :)

**Comment by 66zhaofeng** [comment]
Thank you very much for your help and patient responses.
